crda/remotematcher.go

```diff
@@ -0,0 +1,404 @@
+func WithSource(source string) Option {
+	return func(m *Matcher) error {
+		m.source = source
+		return nil
+	}
+}
+
+// WithRequestConcurrency sets the concurrency limit for the network calls.
+//
+// If not passed to NewMatcher, a defaultRequestConcurrency will be used.
+func WithRequestConcurrency(requestConcurrency int) Option {
+	return func(m *Matcher) error {
+		m.requestConcurrency = requestConcurrency
+		return nil
+	}
+}
+
+// WithBatchSize sets the number of records to be batched per request.
+//
+// If not passed to NewMatcher, a defaultBatchSize will be used.
+func WithBatchSize(batchSize int) Option {
+	return func(m *Matcher) error {
+		m.batchSize = batchSize
+		return nil
+	}
+}
+
+// Name implements driver.Matcher.
+func (*Matcher) Name() string { return "crda" }
+
+// Maps the crda ecosystem to claircore.Repository.Name.
+func ecosystemToRepositoryName(ecosystem string) string {
+	switch ecosystem {
+	case "maven":
+		return "maven"
+	case "pypi":
+		return "pypi"
+	default:
+		panic(fmt.Sprintf("unknown ecosystem %s", ecosystem))
+	}
+}
+
+// Filter implements driver.Matcher.
+func (m *Matcher) Filter(record *claircore.IndexRecord) bool {
+	if record.Repository == nil {
+		return false
+	}
+	return record.Repository.Name == ecosystemToRepositoryName(m.ecosystem)
+}
+
+// Query implements driver.Matcher.
+func (*Matcher) Query() []driver.MatchConstraint {
+	panic("unreachable")
+}
+
+// Vulnerable implements driver.Matcher.
+func (*Matcher) Vulnerable(ctx context.Context, record *claircore.IndexRecord, vuln *claircore.Vulnerability) (bool, error) {
+	// RemoteMatcher can match Package and Vulnerability.
+	panic("unreachable")
+}
+
+// Config is the configuration accepted by the Matcher.
+//
+// By convention, it's in a map key called "crda".
+type Config struct {
+	URL         string `json:"url" yaml:"url"`
+	Concurrency int    `json:"concurrent_requests" yaml:"concurrent_requests"`
+	Source      string `json:"source" yaml:"source"`
+	Key         string `json:"key" yaml:"key"`
+}
+
+// Configure implements driver.MatcherConfigurable.
+func (m *Matcher) Configure(ctx context.Context, f driver.MatcherConfigUnmarshaler, c *http.Client) error {
+	ctx = baggage.ContextWithValues(ctx,
+		label.String("component", "crda/Matcher.Configure"))
+	var cfg Config
+	if err := f(&cfg); err != nil {
+		return err
+	}
+
+	if cfg.Concurrency > 0 {
+		m.requestConcurrency = cfg.Concurrency
+		zlog.Info(ctx).
+			Msg("configured concurrent requests")
+	}
+	if cfg.URL != "" {
+		u, err := url.Parse(cfg.URL)
+		if err != nil {
+			return err
+		}
+		m.url = u
+		zlog.Info(ctx).
+			Msg("configured API URL")
+	}
+	if cfg.Source != "" {
+		m.source = cfg.Source
+		zlog.Info(ctx).
+			Msg("configured source of clair user")
+	}
+	if cfg.Key != "" {
+		m.key = cfg.Key
+		zlog.Info(ctx).
+			Msg("configured 3-scale-key for crda api")
+	}
+	m.client = c
+	zlog.Info(ctx).
+		Msg("configured HTTP client")
```

Comment:
Reuse/remove messages from above, as these all seem to be the same tunables.

Reply:
ACK

Reply:
Removed messages.